File: spark-summit-2018-London/smile-prj/smile-test/scala-prj/src/main/scala/com/github/skonto/RForestTest.scala
package com.github.skonto

import java.util.concurrent.TimeUnit

import smile.classification._
import smile.data.AttributeDataset
import smile.{data, read}

class RForestTest {
  def runModel(path: String = "src/main/resources/iris.arff"): Unit = {
    val data: AttributeDataset = read.arff(path, 4)
    val (x, y) = data.unzipInt
    val rf = RForestTest.time{ randomForest(x, y) }

    println(s"Error = ${rf.error}")
    (0 to x.size-1).foreach{ i: Int =>
      println(RForestTest.time {
        rf.predict(x(i))
      })
    }
  }
}

object RForestTest {
  def main(args: Array[String]): Unit = {
    val rf = new RForestTest()
    if(args.length == 1){
      rf.runModel(args(0))
    } else {
      rf.runModel()
    }
  }

  def time[R](block: => R): R = {
    val t0 = System.nanoTime()
    val result = block
    val t1 = System.nanoTime()
    println("Elapsed time: " + (t1 - t0)/1000000.0 + "ms")
    result
  }
}

File: spark-summit-2018-London/mltest/feature/src/main/java/com/github/com/ml_test/feature/RForestTest.java
package com.github.com.ml_test.feature;

import android.content.res.AssetManager;

import java.io.FileInputStream;
import java.io.InputStream;

import smile.classification.RandomForest;
import smile.data.AttributeDataset;
import smile.data.parser.ArffParser;

public class RForestTest {
    public static String runModel(InputStream path) {
        ArffParser arffParser = new ArffParser();
        arffParser.setResponseIndex(4);
        String ret = "no value";

        StringBuilder sb = new StringBuilder();

        try {
            AttributeDataset data =
                    arffParser.parse(path);

            double[][] x = data.toArray(new double[data.size()][]);
            int[] y = data.toArray(new int[data.size()]);

            long start = System.nanoTime();
            RandomForest rf = new RandomForest(x,y,500);
            long finish = System.nanoTime();
            long timeElapsed = finish - start;

            sb.append("Time for creating the RF model:" + timeElapsed/1000000.0 + "ms\n");
            sb.append("Model error = " + rf.error() + "\n");
            for (int i = 0; i < x.length -1; i++) {

                start = System.nanoTime();
                int classNum = rf.predict(x[i]);
                finish = System.nanoTime();
                timeElapsed = finish - start;
                sb.append("prediction class: " + classNum + " took " + timeElapsed/1000000.0 + "ms\n");
            }
            ret = sb.toString();

        } catch (java.text.ParseException | java.io.IOException e){
            e.printStackTrace();
        }
        return ret;
    }
}
